### PR TITLE
Develop mask action

### PR DESCRIPTION
**Pull request by vincentpierre**
Do not merge : POC only
 - [ ] Make sure the Way to specify masks makes sense 
 - [ ] Make sure the trainer/Model implementation does not slow down training
 - [ ] Reimplement for BC
 - [ ] Ensure the new proto file makes sense
 - [ ] Provide example environment to validate method.

### CODE REVIEW

**Comment by awjuliani** [comment]
@xiaomaogy Was going to work on this. Oh well 😆 . Comments coming soon.

File: python/unityagents/brain.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 class BrainInfo:
     def __init__(self, visual_observation, vector_observation, text_observations, memory=None,
                  reward=None, agents=None, local_done=None,
-                 vector_action=None, text_action=None, max_reached=None):
+                 vector_action=None, text_action=None, max_reached=None, masked_action=None):
```
Comment (awjuliani) [inline python/unityagents/brain.py]: We should probably call this `action_mask` rather than `masked_action`.

File: python/unityagents/brain.py [hunk 2]
```diff
@@ -18,6 +18,7 @@ def __init__(self, visual_observation, vector_observation, text_observations, me
         self.agents = agents
         self.previous_vector_actions = vector_action
         self.previous_text_actions = text_action
+        self.masked_actions = masked_action
```
Comment (awjuliani) [inline python/unityagents/brain.py]: Same here: `action_mask`

File: python/unityagents/environment.py [hunk 3]
```diff
@@ -457,6 +457,11 @@ def _get_state(self, output: UnityRLOutput) -> (AllBrainInfo, bool):
             else:
                 [x.memories.extend([0] * (memory_size - len(x.memories))) for x in agent_info_list]
                 memory = np.array([x.memories for x in agent_info_list])
+            m_actions = np.ones((len(agent_info_list), self.brains[b].vector_action_space_size))
```
Comment (awjuliani) [inline python/unityagents/environment.py]: Can you explain why you apply the mask here?

File: python/unitytrainers/models.py [hunk 4]
```diff
@@ -225,8 +225,15 @@ def create_dc_actor_critic(self, h_size, num_layers):
         self.policy = tf.layers.dense(hidden, self.a_size, activation=None, use_bias=False,
                                       kernel_initializer=c_layers.variance_scaling_initializer(factor=0.01))
 
-        self.all_probs = tf.nn.softmax(self.policy, name="action_probs")
-        output = tf.multinomial(self.policy, 1)
+        self.masked_actions = tf.placeholder(shape=[None, self.a_size], dtype=tf.float32, name="masked_actions")
```
Comment (awjuliani) [inline python/unitytrainers/models.py]: `self.action_mask`

File: unity-environment/Assets/ML-Agents/Scripts/Agent.cs [hunk 5]
```diff
@@ -583,6 +591,23 @@ public virtual void CollectObservations()
 
         }
 
+        protected void MaskDiscreteAction(bool[] mask)
+        {
+            if (brain.brainParameters.vectorActionSpaceType == SpaceType.continuous)
+            {
+                throw new UnityAgentsException(
+                    "Invalid : Cannot set a mask for actions with Continuous Control.");
```
Comment (awjuliani) [inline unity-environment/Assets/ML-Agents/Scripts/Agent.cs]: This error should maybe be something like "Cannot use action mask with continuous action space."

File: unity-environment/Assets/ML-Agents/Scripts/Agent.cs [hunk 6]
```diff
@@ -583,6 +591,23 @@ public virtual void CollectObservations()
 
         }
 
+        protected void MaskDiscreteAction(bool[] mask)
```
Comment (awjuliani) [inline unity-environment/Assets/ML-Agents/Scripts/Agent.cs]: Maybe call this `SetActionMask()`?

Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Scripts/Agent.cs]: It should also have an overload that takes ints of 1/0.

**Comment by awjuliani** [comment]
Overall it looks good. I made some comments concerning variable and function naming, but nothing major. The actual major work left to do is around documentation, as well as creating an environment which shows off the value of the feature properly (instead of hacked PushBlock).

File: unity-environment/Assets/ML-Agents/Scripts/Agent.cs (hunk 5 above)
Comment (xiaomaogy) [reply unity-environment/Assets/ML-Agents/Scripts/Agent.cs]: Maybe the mask can also apply to continuous action space? For example we could limit the action to be only positive (a mask of (0 - positive infinity)), so that in 3dball it will only turn in a certain direction given certain state.

File: python/unityagents/environment.py (hunk 3 above)
Comment (vincentpierre) [reply python/unityagents/environment.py]: This is where we create the mask. We initialize it with ones and then we add the zeros when appropriate. This is to avoid sending a bunch of ones when there is no mask.

File: python/unityagents/environment.py [hunk 7]
```diff
@@ -457,6 +457,11 @@ def _get_state(self, output: UnityRLOutput) -> (AllBrainInfo, bool):
...
+            for i, x in enumerate(agent_info_list):
```
Comment (mmattar) [inline python/unityagents/environment.py]: better variable names than `i` and `x`

File: python/unityagents/environment.py [hunk 8]
```diff
@@ -457,6 +457,11 @@ def _get_state(self, output: UnityRLOutput) -> (AllBrainInfo, bool):
...
+                if x.masked_actions is not None:
+                    for k in range(len(x.masked_actions)):
```
Comment (mmattar) [inline python/unityagents/environment.py]: - should check for length compatibility and throw a helpful error msg if invalid
- can we use array ops to remove the loop?

File: unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs [hunk 9]
```diff
@@ -221,6 +228,29 @@ public void DecideAction(Dictionary<Agent, AgentInfo> agentInfo)
...
+                        }
+                        else
+                        {
+                            maskedActions[i, j] = 1.0f;
+                        }
+                    }
+
+                    i++;
+                }
+
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs]: remove empty line

File: unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs [hunk 10]
```diff
@@ -221,6 +228,29 @@ public void DecideAction(Dictionary<Agent, AgentInfo> agentInfo)
...
+                        if (agentInfo[agent].actionMasks != null)
+                        {
+                            maskedActions[i, j] = agentInfo[agent].actionMasks[j] ? 1.0f : 0.0f;
+                        }
+                        else
+                        {
+                            maskedActions[i, j] = 1.0f;
+                        }
+                    }
+
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs]: remove empty line

File: unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs [hunk 11]
```diff
@@ -158,6 +160,11 @@ public void InitializeCoreBrain(MLAgents.Batcher brainBatcher)
                 {
                     hasPrevAction = true;
                 }
+
+                if (graph[graphScope + "masked_actions"] != null)
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs]: - "masked_scope" should be a static variable.
- replace with: `hasMaskedActions = graph[graphScope + "masked_actions"] != null`

File: unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs [hunk 12]
```diff
@@ -310,6 +340,12 @@ public void DecideAction(Dictionary<Agent, AgentInfo> agentInfo)
             {
                 runner.AddInput(graph[graphScope + PreviousActionPlaceholderName][0], inputPrevAction);
             }
+            
+            // Create the mask action tensor
+            if (hasMaskedActions)
+            {
+                runner.AddInput(graph[graphScope + "masked_actions"][0], maskedActions);
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs]: use static variable that references "masked_actions"

File: python/unitytrainers/models.py [hunk 13]
```diff
@@ -225,8 +225,15 @@ def create_dc_actor_critic(self, h_size, num_layers):
         self.policy = tf.layers.dense(hidden, self.a_size, activation=None, use_bias=False,
                                       kernel_initializer=c_layers.variance_scaling_initializer(factor=0.01))
 
-        self.all_probs = tf.nn.softmax(self.policy, name="action_probs")
-        output = tf.multinomial(self.policy, 1)
+        self.action_masks = tf.placeholder(shape=[None, self.a_size], dtype=tf.float32, name="action_masks")
```
Comment (mmattar) [inline python/unitytrainers/models.py]: `action_masks` should be a global var

File: python/unitytrainers/models.py [hunk 14]
```diff
@@ -225,8 +225,15 @@ def create_dc_actor_critic(self, h_size, num_layers):
...
+        raw_probs = tf.nn.softmax(self.policy, name="action_probs")
```
Comment (mmattar) [inline python/unitytrainers/models.py]: `action_probs` should be a global var

File: python/unitytrainers/ppo/trainer.py [hunk 15]
```diff
@@ -291,6 +293,8 @@ def add_experiences(self, curr_all_info: AllBrainInfo, next_all_info: AllBrainIn
                     if self.is_continuous_action:
                         actions_pre = stored_take_action_outputs[self.model.output_pre]
                         self.training_buffer[agent_id]['actions_pre'].append(actions_pre[idx])
+                    else:
+                        self.training_buffer[agent_id]['action_masks'].append(stored_info.masked_actions[idx])
```
Comment (mmattar) [inline python/unitytrainers/ppo/trainer.py]: make `action_masks` a global var

File: python/unitytrainers/ppo/trainer.py [hunk 16]
```diff
@@ -421,6 +425,9 @@ def update_model(self):
                     feed_dict[self.model.action_holder] = np.array(buffer['actions'][start:end]).flatten()
                     if self.use_recurrent:
                         feed_dict[self.model.prev_action] = np.array(buffer['prev_action'][start:end]).flatten()
+                    feed_dict[self.model.action_masks] = np.array(buffer['action_masks'][start:end]).reshape(
```
Comment (mmattar) [inline python/unitytrainers/ppo/trainer.py]: make `action_masks` a global var

File: unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs [hunk 17]
```diff
@@ -221,6 +228,29 @@ public void DecideAction(Dictionary<Agent, AgentInfo> agentInfo)
                 }
             }
 
+            if (hasMaskedActions)
+            {
+                maskedActions = new float[currentBatchSize, brain.brainParameters.vectorActionSize];
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs]: can we initialize once and re-use?

File: unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs (hunk 11 above)
Comment (vincentpierre) [reply unity-environment/Assets/ML-Agents/Scripts/CoreBrainInternal.cs]: Can you explain the first bullet? A static variable would pose problems if it is different for multiple Internal Brains. Am I missing something?

File: python/unitytrainers/models.py (hunk 13 above)
Comment (vincentpierre) [reply python/unitytrainers/models.py]: Why is that ? Shouldn't it be the same as `policy` ?
